packages/console/src/pages/SignInExperience/tabs/SignUpAndSignInTab/components/SignInMethodEditBox/index.tsx

```diff
@@ -0,0 +1,190 @@
+} from './utilities';
+
+type Props = {
+  value: SignInMethod[];
+  onChange: (value: SignInMethod[]) => void;
+  requiredSignInIdentifiers: SignInIdentifier[];
+  isPasswordRequired: boolean;
+  isVerificationRequired: boolean;
+};
+
+const SignInMethodEditBox = ({
+  value,
+  onChange,
+  requiredSignInIdentifiers,
+  isPasswordRequired,
+  isVerificationRequired,
+}: Props) => {
+  const { t } = useTranslation(undefined, { keyPrefix: 'admin_console' });
+
+  const signInIdentifierOptions = signInIdentifiers.filter((candidateIdentifier) =>
+    value.every(({ identifier }) => identifier !== candidateIdentifier)
+  );
+
+  // Note: add a reference to avoid infinite loop when change the value by `useEffect`
+  const signInMethods = useRef(value);
+
+  const handleChange = useCallback(
+    (value: SignInMethod[]) => {
+      // eslint-disable-next-line @silverhand/fp/no-mutation
+      signInMethods.current = value;
+      onChange(value);
+    },
+    [onChange]
+  );
+
+  const addSignInMethod = useCallback(
+    (identifier: SignInIdentifier) => {
+      handleChange(
+        appendSignInMethodIfNotExist(value, identifier, isPasswordRequired, isVerificationRequired)
+      );
+    },
+    [handleChange, value, isPasswordRequired, isVerificationRequired]
+  );
+
+  useEffect(() => {
+    handleChange(
+      requiredSignInIdentifiers
+        .reduce(
+          (previous, current) =>
+            appendSignInMethodIfNotExist(
+              previous,
+              current,
+              isPasswordRequired,
+              isVerificationRequired
+            ),
+          signInMethods.current
+        )
+        .map((method) => ({
+          ...method,
+          password: isPasswordRequired,
+          verificationCode: isVerificationRequired,
+        }))
+    );
+  }, [handleChange, isPasswordRequired, isVerificationRequired, requiredSignInIdentifiers]);
+
+  const removeSignInMethod = (identifier: SignInIdentifier) => {
+    handleChange(value.filter((method) => method.identifier !== identifier));
+  };
+
+  const onMoveItem = (dragIndex: number, hoverIndex: number) => {
+    const dragItem = value[dragIndex];
+    const hoverItem = value[hoverIndex];
+
+    if (!dragItem || !hoverItem) {
+      return;
+    }
+
+    handleChange(
+      value.map((value_, index) => {
+        if (index === dragIndex) {
+          return hoverItem;
+        }
+
+        if (index === hoverIndex) {
+          return dragItem;
+        }
+
+        return value_;
+      })
+    );
+  };
+
+  const isDeleteItemDisabled = (signInIdentifier: SignInIdentifier) =>
+    requiredSignInIdentifiers.includes(signInIdentifier);
+
+  return (
+    <div>
+      <DragDropProvider>
+        {value.map(({ identifier, isPasswordPrimary, password, verificationCode }, index) => (
+          <DraggableItem key={identifier} id={identifier} sortIndex={index} moveItem={onMoveItem}>
+            <div key={snakeCase(identifier)} className={styles.signInMethodItem}>
```

Comment:
better extract this div into a component to make it more clear

Reply:
updated.